pkg/provision/storage/perWorkspaceStorage_test.go

```diff
@@ -0,0 +1,97 @@
+//
+// Copyright (c) 2019-2022 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package storage
+
+import (
+	"testing"
+
+	dw "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	"github.com/devfile/devworkspace-operator/pkg/provision/sync"
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	"github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+)
+
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(v1alpha1.AddToScheme(scheme))
+	utilruntime.Must(dw.AddToScheme(scheme))
+	config.SetConfigForTesting(nil)
+}
+
+func TestRewriteContainerVolumeMountsForPerWorkspaceStorageClass(t *testing.T) {
+	tests := loadAllTestCasesOrPanic(t, "testdata/perWorkspace-storage")
+	setupControllerCfg()
+	perWorkspaceStorage := perWorkspaceStorageProvisioner{}
+
+	for _, tt := range tests {
+		t.Run(tt.Name, func(t *testing.T) {
+			// sanity check that file is read correctly.
+			assert.NotNil(t, tt.Input.Workspace, "Input does not define workspace")
+			workspace := &dw.DevWorkspace{}
+			workspace.Spec.Template = *tt.Input.Workspace
+			workspace.Status.DevWorkspaceId = tt.Input.DevWorkspaceID
+			workspace.Namespace = "test-namespace"
+
+			workspacePVC, err := getPerWorkspacePVCSpec(workspace, scheme, "10Gi")
+
+			if err != nil {
+				t.Fatalf("Failure during setup: %s", err)
+			}
+			workspacePVC.Status.Phase = corev1.ClaimBound
```

Comment:
It's unclear why this is needed, and why the workspacePVC should already always exist on the cluster before calling provisionStorage. The better approach may be to call `ProvisionStorage` twice in the test (first to create the PVC, second to rewrite container VMs) and verify that each step has the expected output (error the first time, nil error the second, etc.)

Reply:
I agree -  `ProvisionStorage` should be good to create the PVC rather than creating it on the cluster before calling `ProvisionStorage`.

However, in my testing, it seems that for certain cases,  `ProvisionStorage` will return early because a PVC is not required (https://github.com/devfile/devworkspace-operator/pull/818/files#diff-a3fc5cfc1fe497d9039f7540c20597bc9e87b4ebbe9ab58d6bc25b84dc647ed6R115)

I'm not sure if I should:
- Remove those test cases (most likely not)
- Skip the check for PVC creation, and owner reference check for these test cases
- Or something else that I haven't considered yet.

I'm going to submit the updated PR, so you can see what I mean